script/transform-approved-books-data.ts

```diff
@@ -0,0 +1,168 @@
+    && Array.isArray(something.books)
+    && something.books.every(isBookData);
+};
+
+interface ApprovedCollection extends ApprovedBooksData {
+  collection_id: string;
+  server: string;
+}
+
+const isApprovedCollection = (something: any): something is ApprovedCollection => {
+  return typeof something.collection_id === 'string'
+    && typeof something.server === 'string'
+    && isApprovedBooksData(something);
+};
+
+interface ApprovedRepo extends ApprovedBooksData {
+  repo: string;
+}
+
+const isApprovedRepo = (something: any): something is ApprovedRepo => {
+  return typeof something.repo === 'string' && isApprovedBooksData(something);
+};
+
+interface ApprovedVersion {
+  content_version: string;
+  min_code_version: string;
+}
+
+const isApprovedVersion = (something: any): something is ApprovedVersion => {
+  return typeof something.content_version === 'string' && typeof something.min_code_version === 'string';
+};
+
+interface ApprovedVersionCollection extends ApprovedVersion {
+  collection_id: string;
+}
+
+const isApprovedVersionCollection = (something: any): something is ApprovedVersionCollection => {
+  return typeof something.collection_id === 'string' && isApprovedVersion(something);
+};
+
+interface ApprovedVersionRepo extends ApprovedVersion {
+  repo: string;
+}
+
+const isApprovedVersionRepo = (something: any): something is ApprovedVersionRepo => {
+  return typeof something.repo === 'string' && isApprovedVersion(something);
+};
+
+interface ApprovedBooksAndVersions {
+  approved_books: Array<ApprovedCollection | ApprovedRepo>;
+  approved_versions: Array<ApprovedVersionCollection | ApprovedVersionRepo>;
+}
+
+const isApprovedBooksAndVersions = (something: any): something is ApprovedBooksAndVersions => {
+  return Array.isArray(something.approved_books)
+    && something.approved_books.every(
+      (element: any) => isApprovedCollection(element) || isApprovedRepo(element))
+    && something.approved_versions.every(
+      (element: any) => isApprovedVersionCollection(element) || isApprovedVersionRepo(element));
+};
+
+const matchRepoVersion = (repoData: ApprovedRepo) =>
+  (versionData: ApprovedVersionCollection | ApprovedVersionRepo): versionData is ApprovedVersionRepo => {
+  return isApprovedVersionRepo(versionData) && versionData.repo === repoData.repo;
+};
+
+const matchCollectionVersion = (collectionData: ApprovedCollection) =>
+  (versionData: ApprovedVersionCollection | ApprovedVersionRepo): versionData is ApprovedVersionCollection => {
+  return isApprovedVersionCollection(versionData) && versionData.collection_id === collectionData.collection_id;
+};
+
+const getDesiredVersion = (
+  approvedVersions: Array<ApprovedVersionCollection | ApprovedVersionRepo>,
+  colOrRepo: ApprovedRepo | ApprovedCollection
+): string | undefined => {
+  const isCollection = isApprovedCollection(colOrRepo);
+
+  const filter = isCollection
+    ? matchCollectionVersion(colOrRepo as ApprovedCollection)
```

Comment:
you have a guard on line 97, you shouldn't have to cast (i mean you do here, but you should be able to restructure and not)